snapcraft/internal/lifecycle/_status_cache.py

```diff
@@ -0,0 +1,174 @@
+                         step: steps.Step) -> pluginhandler.DirtyReport:
+        """Obtain the dirty report for a given step of the given part.
+
+        :param pluginhandler.PluginHandler part: Part in question.
+        :param steps.Step step: Step in question.
+        :return: Dirty report (could be None)
+        :rtype: pluginhandler.DirtyReport
+        """
+        self._ensure_dirty_report(part, step)
+        return self._dirty_reports[part.name][step]
+
+    def clear_step(self, part: pluginhandler.PluginHandler,
+                   step: steps.Step) -> None:
+        """Clear the given step of the given part from the cache.
+
+        :param pluginhandler.PluginHandler part: Part in question.
+        :param steps.Step step: Step in question.
+
+        This function does nothing if the step wasn't cached.
+        """
+        if part.name in self._steps_run:
+            _remove_key(self._steps_run[part.name], step)
+            if not self._steps_run[part.name]:
+                _del_key(self._steps_run, part.name)
+        _del_key(self._dirty_reports[part.name], step)
+        if not self._dirty_reports[part.name]:
+            _del_key(self._dirty_reports, part.name)
+
+    def _ensure_steps_run(self, part: pluginhandler.PluginHandler) -> None:
+        if part.name not in self._steps_run:
+            self._steps_run[part.name] = _get_steps_run(part)
+
+    def _ensure_dirty_report(self, part: pluginhandler.PluginHandler,
+                             step: steps.Step) -> None:
+        if step not in self._dirty_reports[part.name]:
+            self._dirty_reports[part.name][step] = part.get_dirty_report(step)
+
+            # The dirty report from the PluginHandler only takes into account
+            # properties specific to that part. We need to expand that here to
+            # also take its dependencies (if any) into account, but only if
+            # it's not already dirty.
+            if not self._dirty_reports[part.name][step]:
+                dependencies = self.config.parts.get_dependencies(
+                    part.name, recursive=True)
+                prerequisite_step = steps.get_dependency_prerequisite_step(
+                    step)
+                changed_dependencies = []
+                with contextlib.suppress(errors.StepHasNotRunError):
+                    step_timestamp = part.step_timestamp(step)
+                    for dependency in dependencies:
+                        # Make sure the prerequisite step of this dependency
+                        # has not run more recently than (or should run
+                        # _before_) this step.
+                        try:
+                            prerequisite_timestamp = dependency.step_timestamp(
+                                prerequisite_step)
+                        except errors.StepHasNotRunError:
+                            dependency_changed = True
+                        else:
```

Comment:
I always have to lookup if `else` is for the `try` succeeding or no matching `except` :man_shrugging: